bin/start-func-test-server.js

```diff
@@ -0,0 +1,191 @@
+/* @flow */
+/* eslint-disable no-console */
+// This starts a docker server for functional tests to access.
+// The script waits for the server to start and prints some logs
+// to help with debugging.
+
+const fs = require('fs');
+const path = require('path');
+const childProcess = require('child_process');
+
+const root = path.join(__dirname, '..');
+console.log(`Working directory: ${root}`);
+const containerIdFile = path.join(root, 'docker-container-id.txt');
+
+// TODO: make these configurable when we need to start
+// servers for multiple apps.
+const appInstance = 'disco';
+const nodeEnv = 'uitests';
+
+function logDivider(heading) {
+  console.log(`${'='.repeat(35)} ${heading} ${'='.repeat(35)}`);
+}
+
+function shell(cmd, args) {
+  // Execs a command as if it were part of the parent shell
+  // (i.e. the output is unbuffered and is displayed in the console).
+  logDivider('BEGIN shell');
+  const cmdString = `${cmd} ${args.join(' ')}`;
+  console.log(`Shell: ${cmdString}`);
+  return new Promise((resolve, reject) => {
+    const child = childProcess.spawn(cmd, args, {
+      cwd: root,
+    });
+
+    child.stdout.on('data', (data) => {
+      process.stdout.write(data.toString());
+    });
+    child.stdout.on('error', (error) => reject(error));
+
+    child.stderr.on('data', (data) => {
+      process.stderr.write(data.toString());
+    });
+    child.stderr.on('error', (error) => reject(error));
+
+    child.on('error', (error) => reject(error));
+    child.on('exit', (exitCode) => {
+      logDivider('END shell');
+      if (exitCode !== 0) {
+        reject(new Error(
+          `shell command failed: ${cmdString} (exit: ${exitCode || '[empty]'})`));
+      }
+      resolve();
+    });
+  });
+}
+
+function exec(cmd, argParts, { quiet = false } = {}) {
+  const cmdString = `${cmd} ${argParts.join(' ')}`;
+  if (!quiet) {
+    logDivider('BEGIN exec');
+    console.log(`Exec: ${cmdString}`);
+  }
+  return new Promise((resolve, reject) => {
+    childProcess.exec(cmdString, { cwd: root }, (error, stdout, stderr) => {
+      let gotOutput = false;
```

Comment:
Personally I found `gotOutput` a bit of an odd variable name to grok at first. Maybe `receivedOutput` would be better?